keps/sig-autoscaling/5325-hpa-pod-selection-accuracy/README.md

```diff
@@ -0,0 +1,879 @@
+
+### Risks and Mitigations
+
+* Backward compatibility: Mitigated by making the new behavior opt-in with the current behavior as default.
+* User confusion: We'll provide clear documentation on when and how to use each strategy.
+
+## Design Details
+
+The HPA specification (v2) will be extended with a new boolean field:
+```yaml
+apiVersion: autoscaling/v2
+kind: HorizontalPodAutoscaler
+spec:
+  # Existing fields...
+  strictPodSelection: true  # Default: false
+```
+When the HPA controller processes an HPA resource:
+
+* If strictPodSelection is not specified or set to false:
+  * The controller will use the current behavior, selecting all pods that match the target workload's label selector
+  * This maintains backward compatibility with existing HPAs
+
+* If strictPodSelection is set to true:
+  * The controller will identify the target workload (e.g., Deployment)
+  * It will traverse the ownership chain (e.g., Deployment → ReplicaSet → Pods)
+  * Only pods that are directly owned by the target workload through this chain will be included in metric collection
+  * Pods that match labels but aren't in the ownership chain will be excluded
+
+Additionally, the HPA status will be enhanced to include information about the pod selection:
+```yaml
+status:
+  # Existing fields...
+  podSelectionInfo:
+    strategy: "Strict"  # or "Label" when strictPodSelection is false
+```
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+None required.
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
```

Comment:
You need to write the planned tests and list all the packages you're planning to touch with their current coverage.

Reply:
```suggestion
- `pkg/apis/autoscaling/v1`: `2025-06-01` - `28.6%`
- `pkg/apis/autoscaling/v2`: `2025-06-01` - `75.6%`
- `pkg/apis/autoscaling/v2beta1`: `2025-06-01` - `69.1%`
- `pkg/apis/autoscaling/v2beta2`: `2025-06-01` - `82.1%`
- `pkg/controller/podautoscaler`: `2025-06-01` - `88.3%`
```